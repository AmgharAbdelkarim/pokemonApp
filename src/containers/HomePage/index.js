import React, { useEffect } from "react";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import { ContentWrapper } from "containers/HomePage/style";
import Pokemon from "components/Box";
import Button from "components/Button";
import LoadingState from "components/Loading";
import Error from "components/Error";
import EmptyState from "components/EmptyState";

import {
  getPokemonSelector,
  getSizeSelector,
  getTypesSelector,
  getLoadingSelector,
  getHasErrorSelector,
  getCurrentUrlSelector,
  getNextUrlSelector,
  getPrevUrlSelector,
  getSelectSelectedTypes,
} from "store/selectors";
import {
  getPokemon,
  setTypeSelected,
  setCurrentUrl,
  setSize,
  getTypes,
  clearTypeSelected,
} from "store/action";

const Home = ({
  pokemon,
  size,
  types,
  loading,
  hasError,
  currentUrl,
  nextUrl,
  prevUrl,
  getPokemon,
  selectedTypes,
  setTypeSelected,
  setCurrentUrl,
  setSize,
  getTypes,
  clearTypeSelected,
  history,
}) => {
  useEffect(() => {
    getPokemon();
  }, [currentUrl]);

  useEffect(() => {
    getTypes();
  }, []);

  const nextHandler = () => {
    if (!nextUrl) return setSize(size + 20);
    setCurrentUrl(nextUrl);
  };

  const prevHandler = () => {
    if (!nextUrl) return setSize(size - 20);
    setCurrentUrl(prevUrl);
  };

  const filterByType = (currentUrl, name) => {
    setCurrentUrl(currentUrl);
    setTypeSelected(name);
  };

  return (
    <ContentWrapper id="content">
      {loading ? (
        <LoadingState />
      ) : hasError ? (
        <Error />
      ) : pokemon.length > 0 ? (
        <React.Fragment>
          <div className="filter_box">
            <h5>filter Pokemon By Type</h5>
            {selectedTypes.length > 0 && (
              <Button
                className="clear_button"
                clickHandler={() => {
                  setCurrentUrl("https://pokeapi.co/api/v2/pokemon");
                  clearTypeSelected();
                }}
              >
                Clear
              </Button>
            )}
            <ul>
              {types.map(({ name, url }) => (
                <li
                  onClick={() => filterByType(url, name)}
                  aria-hidden="true"
                  key={name}
                  className={selectedTypes.indexOf(name) >= 0 ? "selected" : ""}
                >
                  <span>{name}</span>
                </li>
              ))}
            </ul>
          </div>
          <div className="item_container">
            <div className="item_wrapper">
              {pokemon.slice(size, size + 20).map((p, index) => (
                <Pokemon
                  key={index}
                  name={p.name || p.pokemon.name}
                  clickHandler={() =>
                    history.push(`/pokemon/${p.name || p.pokemon.name}`)
                  }
                />
              ))}
            </div>
            <div className="button_wrapper">
              <Button
                disabled={prevUrl || size > 19 ? false : true}
                clickHandler={() => prevHandler()}
              >
                previous
              </Button>
              <Button
                disabled={nextUrl || size < pokemon.length - 19 ? false : true}
                clickHandler={() => nextHandler()}
              >
                next
              </Button>
            </div>
          </div>
        </React.Fragment>
      ) : (
        <EmptyState />
      )}
    </ContentWrapper>
  );
};

const mapStateToProps = createStructuredSelector({
  pokemon: getPokemonSelector,
  size: getSizeSelector,
  types: getTypesSelector,
  loading: getLoadingSelector,
  hasError: getHasErrorSelector,
  currentUrl: getCurrentUrlSelector,
  nextUrl: getNextUrlSelector,
  prevUrl: getPrevUrlSelector,
  selectedTypes: getSelectSelectedTypes,
});

const mapDispatchToProps = {
  getPokemon,
  setTypeSelected,
  setCurrentUrl,
  setSize,
  getTypes,
  clearTypeSelected,
};
export default connect(mapStateToProps, mapDispatchToProps)(Home);

export { Home };
