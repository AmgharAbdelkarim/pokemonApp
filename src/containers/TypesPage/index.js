import React, { useEffect } from "react";
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";

import Pokemon from "components/Box";
import Button from "components/Button";
import { TypeWrapper } from "containers/TypesPage/style";
import LoadingState from "components/Loading";
import Error from "components/Error";
import EmptyState from "components/EmptyState";

import {
  getPokemonTypesSelector,
  getSizeSelector,
  getLoadingSelector,
  getHasErrorSelector,
} from "store/selectors";
import { getPokemonTypes, setSize } from "store/action";

const Types = ({
  match,
  pokemon,
  size,
  loading,
  hasError,
  getPokemonTypes,
  setSize,
  history,
}) => {
  const { params } = match;

  useEffect(() => {
    getPokemonTypes(params.type);
  }, []);
  const nextHandler = () => {
    setSize(size + 20);
  };
  const prevHandler = () => {
    setSize(size - 20);
  };
  return (
    <TypeWrapper id="content">
      {loading ? (
        <LoadingState />
      ) : hasError ? (
        <Error />
      ) : pokemon.length > 0 ? (
        <React.Fragment>
          <div className="item_wrapper">
            {pokemon.slice(size, size + 20).map(({ pokemon }, index) => (
              <Pokemon
                clickHandler={() => history.push(`/pokemon/${pokemon.name}`)}
                name={pokemon.name}
                key={index}
              />
            ))}
          </div>
          <div className="button_wrapper">
            <Button
              disabled={size > 19 ? false : true}
              clickHandler={() => prevHandler()}
            >
              previous
            </Button>
            <Button
              disabled={size < pokemon.length - 19 ? false : true}
              clickHandler={() => nextHandler()}
            >
              next
            </Button>
          </div>
        </React.Fragment>
      ) : (
        <EmptyState />
      )}
    </TypeWrapper>
  );
};

const mapStateToProps = createStructuredSelector({
  pokemon: getPokemonTypesSelector,
  size: getSizeSelector,
  loading: getLoadingSelector,
  hasError: getHasErrorSelector,
});

const mapDispatchToProps = {
  getPokemonTypes,
  setSize,
};

export default connect(mapStateToProps, mapDispatchToProps)(Types);

export { Types };
