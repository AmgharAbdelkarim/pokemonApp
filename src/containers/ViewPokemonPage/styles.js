import styled from 'styled-components';

export const PokemonContainer = styled.div`
  margin: 0 auto;
  padding: 10px 0px;
`;
