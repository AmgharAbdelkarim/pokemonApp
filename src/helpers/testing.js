import React from "react";
import { Provider } from "react-redux";
import { render } from "@testing-library/react";
import store from "store";
import { Router as Switch } from "react-router-dom";

import { createBrowserHistory } from "history";

export const history = createBrowserHistory();

export const Wrapper = (props) => {
  return (
    <Provider store={store()}>
      <Switch history={history}>{props.children}</Switch>
    </Provider>
  );
};

export const renderWithRedux = (ui, renderOptions = {}) => {
  return {
    ...render(ui, {
      wrapper: Wrapper,
      ...renderOptions,
    }),
    store,
  };
};
