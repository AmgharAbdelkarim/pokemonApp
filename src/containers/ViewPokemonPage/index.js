import React, { useEffect } from "react";

import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";

import Card from "components/Card";
import { PokemonContainer } from "containers/ViewPokemonPage/styles";
import Box from "components/Box";
import LoadingState from "components/Loading";
import Error from "components/Error";
import EmptyState from "components/EmptyState";

import {
  getPokemonDetailsSelector,
  getLoadingSelector,
  getHasErrorSelector,
  getPokemonEvolutionSelector,
} from "store/selectors";
import { getPokemonDetails, getPokemonEvolutions } from "store/action";

const ViewPokemon = ({
  match,
  history,
  pokemon,
  getPokemonDetails,
  getPokemonEvolutions,
  evolution,
  loading,
  hasError,
}) => {
  const { params } = match;

  useEffect(() => {
    getPokemonDetails(params.name);
  }, [params.name]);
  useEffect(() => {
    getPokemonEvolutions(pokemon.id);
  }, [pokemon]);
  return (
    <PokemonContainer id="content">
      {loading ? (
        <LoadingState />
      ) : hasError ? (
        <Error />
      ) : pokemon.name ? (
        <React.Fragment>
          <Card pokemon={pokemon} clickHandler={(url) => history.push(url)} />
          {evolution.length > 0 && (
            <React.Fragment>
              <h3>pokemon evolutions</h3>
              {evolution.map(({ species }, index) => (
                <Box
                  key={index}
                  name={species.name}
                  clickHandler={() => history.push(`/pokemon/${species.name}`)}
                />
              ))}
            </React.Fragment>
          )}
        </React.Fragment>
      ) : (
        <EmptyState />
      )}
    </PokemonContainer>
  );
};

const mapStateToProps = createStructuredSelector({
  pokemon: getPokemonDetailsSelector,
  loading: getLoadingSelector,
  hasError: getHasErrorSelector,
  evolution: getPokemonEvolutionSelector,
});

const mapDispatchToProps = {
  getPokemonDetails,
  getPokemonEvolutions,
};

export default connect(mapStateToProps, mapDispatchToProps)(ViewPokemon);
export { ViewPokemon };
