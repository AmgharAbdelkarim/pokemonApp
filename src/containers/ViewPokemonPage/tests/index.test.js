import { cleanup, fireEvent, waitFor } from "@testing-library/react";

import { renderWithRedux } from "helpers/testing";

import { ViewPokemon } from "containers/ViewPokemonPage";
import { createBrowserHistory } from "history";

afterEach(cleanup);

describe("ViewPokemon", () => {
  let props;
  beforeEach(() => {
    props = {
      pokemon: {
        name: "fake name",
        weight: 124,
        height: 12,
        abilities: [
          {
            ability: {
              name: "water-absorb",
              url: "https://pokeapi.co/api/v2/ability/11/",
            },
            is_hidden: false,
            slot: 1,
          },
          {
            ability: {
              name: "damp",
              url: "https://pokeapi.co/api/v2/ability/6/",
            },
            is_hidden: false,
            slot: 2,
          },
          {
            ability: {
              name: "swift-swim",
              url: "https://pokeapi.co/api/v2/ability/33/",
            },
            is_hidden: true,
            slot: 3,
          },
        ],
        types: [
          {
            type: {
              name: "fakeType",
            },
          },
        ],
        sprites: {
          front_default: "fake image Url",
        },
      },
      evolution: [
        {
          species: {
            name: "fakeName",
          },
        },
      ],
      loading: false,
      hasError: false,
      getPokemonDetails: jest.fn(),
      getPokemonEvolutions: jest.fn(),
      match: {
        params: "fakePokemon",
      },
    };
  });
  afterEach(cleanup);
  it("should ViewPokemon render correctly", () => {
    const { asFragment } = renderWithRedux(<ViewPokemon {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
  it("should render loading state if loading is true", () => {
    let newProps = { ...props, loading: true };
    const { getByText } = renderWithRedux(<ViewPokemon {...newProps} />);
    expect(getByText("Loading...")).toBeInTheDocument();
  });
  it("should render error state if hasError is true", () => {
    let newProps = { ...props, hasError: true };
    const { getByText } = renderWithRedux(<ViewPokemon {...newProps} />);
    expect(getByText("Something Went Wrong")).toBeInTheDocument();
  });
  it("should redirect to view pokemon page if i click on pokemon", async () => {
    const history = createBrowserHistory();
    jest.spyOn(history, "push");
    const { getByText } = renderWithRedux(
      <ViewPokemon {...props} history={history} />
    );
    const pokemon = getByText("fakeName");
    fireEvent.click(pokemon);
    await waitFor(() =>
      expect(history.push).toHaveBeenCalledWith("/pokemon/fakeName")
    );
  });
  it("should redirect to view type page if i click on type", async () => {
    const history = createBrowserHistory();
    jest.spyOn(history, "push");
    const { getByText } = renderWithRedux(
      <ViewPokemon {...props} history={history} />
    );
    const pokemon = getByText("fakeType");
    fireEvent.click(pokemon);
    await waitFor(() =>
      expect(history.push).toHaveBeenCalledWith("/type/fakeType")
    );
  });
});
