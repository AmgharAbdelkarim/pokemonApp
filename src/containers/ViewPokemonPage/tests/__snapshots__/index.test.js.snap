// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewPokemon should ViewPokemon render correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-gsTCUz jQNeIX"
    id="content"
  >
    <div
      class="sc-bdfBwQ kWMDsT"
    >
      <div
        class="Card_img"
      >
        <img
          alt=""
          src="fake image Url"
        />
      </div>
      <div
        class="Card_name"
      >
        fake name
      </div>
      <div
        class="Card_types"
      >
        <div
          aria-hidden="true"
          class="Card_type"
        >
          fakeType
        </div>
      </div>
      <div
        class="Card_info"
      >
        <div
          class="Card_data Card_data-weight"
        >
          <p
            class="title"
          >
            Weight
          </p>
          <p>
            124
          </p>
        </div>
        <div
          class="Card_data Card_data-weight"
        >
          <p
            class="title"
          >
            Height
          </p>
          <p>
            12
          </p>
        </div>
        <div
          class="Card_data Card_data-ability"
        >
          <p
            class="title"
          >
            Ability
          </p>
          <p>
            water-absorb
          </p>
        </div>
      </div>
    </div>
    <h3>
      pokemon evolutions
    </h3>
    <div
      class="sc-dlfnbm bPfrbJ"
    >
      <span>
        fakeName
      </span>
    </div>
  </div>
</DocumentFragment>
`;
